drivers/periph_common/gpio.c

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 2019 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_periph_gpio
+ * @{
+ *
+ * @file        gpio.c
+ * @brief       Fallback Implementations for platform specific GPIO routines.
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include "periph/gpio.h"
+
+#ifndef GPIO_HAVE_INIT_LEVEL
+
+int gpio_init_low(gpio_t pin)
+{
+    gpio_clear(pin);
+    return gpio_init(pin, GPIO_OUT);
+}
```

Comment:
I'm not sure if the fallback implementation should rely on this to work. I see to alternatives:

1. Introduce a new feature. This implementation could use `weak` symbols, so that platforms actually support `gpio_clear()` or `gpio_set()` being called prior to initialization, this can be explicitly enabled and no code needs to be provided.
2. Initialize the pin first and clear it afterwards. That works reliable in for every platform, only that the quality of the experience of `gpio_init_low()` will not be superb on some platforms

Reply:
I was under the impression that 

```C
gpio_set()
gpio_init()
```

was the canonical way to ensure a GPIO is set in RIOT.
This is used in a few places like that - so this is a direct replacement.

If the platform does not support this pattern, it should implement `gpio_init_low()`/`gpio_init_high()`.
Whether we use weak functions or a define doesn't make a difference, I tried to mirror `PERIPH_TIMER_PROVIDES_SET`.

Reply:
My point was not that I prefer the weak symbol over the define. I would like to have two knobs to fiddle with, one controlling whether `gpio_init_low()` / `gpio_init_high()` is supported at all and one to provide a custom implementation. So in total there should be three possible states:

1. The platform does not support `gpio_init_low()` / `gpio_init_high()` at all
2. The platform does support it using the common implementation (e.g. `gpio_set(); gpio_init()`)
3. The platform does support it using a platform specific implementation

Or alternatively, the fallback implementation should first initialize the pin and then sets/clears it. That is guaranteed to work, but suboptimal.

> was the canonical way to ensure a GPIO is set on init in RIOT.

IMO, this is not true, as the doc doesn't state this. But the GPIO doc is pretty poor anyway. Maybe you can fix this (at least partially) in this PR and just add a precondition (via `@pre`) that `gpio_set()` and `gpio_clear()` must not be called for pins that have not been initialized using one of the `gpio_init*()` functions to one of the output modes. That way `gpio_init_high()` and `gpio_init_low()` would become the canonical way.

(Other issues are IMO what the return value of `gpio_read()` is, if the pin is not configured as input (with or without pulls). E.g. for `GPIO_OUT`, I would intuitively expect it to return `!=0` after `gpio_set()` and `0` after `gpio_clear()`. But for `GPIO_OD_PU` I would expect `gpio_read()` after `gpio_set()` to potentially still return `0`, if the line is externally pulled low. But none of this is documented.)